### ISSUE TITLE
Type hints

### ISSUE THREAD

**Issue by Tinche** [alisaifee/coredis#7]
Hello.

We are currently using aioredis with a ton of internal type stubs for it, but aioredis is kind of in a transitional/crisis period right now. So we would like to switch to aredis/coredis. If I submitted PRs with type annotations for a bunch of methods, would you consider merging them in?

**Comment by alisaifee** [comment]
Hi @Tinche - I'd strongly recommend holding off on spending any time on adding type hints  (or any non bug fix changes) on the state of the code that is in master branch right now (which is approximately == 2.x branch). The `2.x` series that I've sort of frozen is backward compatible with `aredis` and contains some of the pending pull requests/fixes + python 3.10 support.

I'm currently (~85% done) in the midst of a significant refactor which involves some breaking changes, but also has proper type hinting support as a required goal. I hope to  get something in master and have atleast a beta release out in the next couple of days, and hopefully wrap up documentation cleanup next weekend.

The TL;DR of the breaking changes are:
-  Better consistency with redis docs in terms of method argument names & types (and default values)
-  Use of  enums for pure tokens 
-  Stricter return types and  eventually, overloads for methods that can return different types
-  Removal of non essential extras such as cache
-  [Most probably] drop support for python 3.7
-  [Most probably] drop support using both bytes and strings and just defaulting to strings in and out with always decoded responses
  
In most simple cases the apis will not be significantly different from either `redis-py` or `aredis` so you should be able to drop it in to your code base with trivial effort.

Given what I've described above regarding breaking changes - if that would be manageable for your code base, I'd love to get some eyes (other than mine :grin:) on the 3.x candidate.

An example showing a combination of some of these changes:
Before:
```python
   async def georadius(
        self,
        key,
        longitude,
        latitude,
        radius,
        unit=None,
        withdist=False,
        withcoord=False,
        withhash=False,
        count=None,
        sort=None,
        store=None,
        store_dist=None,
        any=False,
    ):
```

After:
```python
    async def georadius(
        self,
        key: str,
        longitude: Union[int, float],
        latitude: Union[int, float],
        radius: Union[int, float],
        unit: Literal[PureToken.M, PureToken.KM, PureToken.FT, PureToken.MI],
        withcoord: Optional[bool] = None,
        withdist: Optional[bool] = None,
        withhash: Optional[bool] = None,
        count: Optional[int] = None,
        any_: Optional[bool] = None,
        order: Optional[Literal[PureToken.ASC, PureToken.DESC]] = None,
        store: Optional[str] = None,
        storedist: Optional[str] = None,
    ) -> Union[int, Tuple[Union[str, GeoSearchResult], ...]]:
```
Where `GeoSearchResult` and `GeoCoordinates` are still tuples but named.: 
```
class GeoCoordinates(NamedTuple):
    latitude: float
    longitude: float


class GeoSearchResult(NamedTuple):
    name: str
    distance: float
    geohash: int
    coordinates: GeoCoordinates
```

**Comment by alisaifee** [comment]
Thanks - that is exactly the kind of feedback I'd love to get. I was toying with making `Redis` use generics for the return type - and you're absolutely right about not needing the union :+1:

**Comment by alisaifee** [comment]
Sorry for the delay but I've unfortunately not had much time in the last few weeks to work on this. Having said that I just noticed a major addition in redis-py itself to start adding async support (https://github.com/redis/redis-py/pull/1899) which is making me seriously rethink the usefulness of this project.

**Comment by Tinche** [comment]
Good to know, however if this is just bringing aioredis 2.0 into redis-py (which is what it looked like the last time I looked), this project can still be useful. aioredis 2.0 was/(probably still) is much slower than aioredis 1.3 or aredis.

**Comment by Andrew-Chen-Wang** [comment]
As maintainer of aioredis and also the person who did the redis py migration, having another option in the meantime is great, especially a maintained one. I would simply be careful around adding features; it gets very hectic very fast. Additionally, one of the other aioredis maintainers is refactoring the asyncio portion to improve performance back to original status, but until then this library is quite useful :)

**Comment by alisaifee** [comment]
This is an interesting and difficult point in the life of redis python, especially for asyncio.
I've not been able to manage even really getting half way done with the refactoring of aredis to a point that I can reason around the code enough to quickly iterate on adding features (since it was written by someone else).

I will probably have some time the next few weeks to see the type hinting support through. The half baked version so far is in https://github.com/alisaifee/coredis/commit/6ba867549e5f5a70a020b8d65a3d14ee03f2341d but I'm indeed starting to feel that this will quickly get out of hand for a single maintainer

**Comment by alisaifee** [comment]
My design goals for the refactoring are:
- Stay as close to the redis command definition (both for command parameters and return types)
- Minimal runtime overhead
- Runtime type safety as an optional opt-in using beartype
- Departure from the implicit response destructuring using global mapping and try to "inline" how a command is parsed and transformed in to a python equivalent
- Full support for RESP3 to minimize the guess work around parsing response types.

**Comment by alisaifee** [comment]
@Tinche perhaps you can look at #20 which includes type hints. From my local testing the performance doesn't diverge from the 2.x branch but a second opinion wouldn't hurt.

**Comment by alisaifee** [comment]
I've merged the current state of type hints / updates to master and made an initial `3.0.0rc1` release.

Specifically related to the type hints you can see the WIP [docs](https://coredis.readthedocs.io/en/latest/api_reference.html#typing) and the [API documentation](https://coredis.readthedocs.io/en/latest/api.html)

**Comment by Tinche** [comment]
@alisaifee Sorry, I was super busy but I might be able to review soon.

**Comment by alisaifee** [comment]
FYI I tagged / released`3.0.0rc2` which contains the [breaking api changes](https://coredis.readthedocs.io/en/latest/release_notes.html#v3-0-0rc2) that I was referring to.
